load_tokenizer from gcs is broken for llama3-sized tokenizers

**Describe the bug**

HF added a hack that broke my hacky monkey patching of their download machinery to support GCS etc. He who lives by the hack dies by the hack

**To Reproduce**

```python
def test_load_tokenizer_gcs():
    # sort of like a gs:// path insasmuch as it uses fsspec machinery
    tokenizer = load_tokenizer("gs://marin-us-central2/checkpoints/tootsie-8b-giraffe-32k-293fef/hf/step-2999")
```


```
tests/test_hf_utils.py:24 (test_load_tokenizer_gcs)
response = <Response [404]>, endpoint_name = None

    def hf_raise_for_status(response: Response, endpoint_name: Optional[str] = None) -> None:
        """
        Internal version of `response.raise_for_status()` that will refine a
        potential HTTPError. Raised exception will be an instance of `HfHubHTTPError`.
    
        This helper is meant to be the unique method to raise_for_status when making a call
        to the Hugging Face Hub.
    
    
        Example:
        ```py
            import requests
            from huggingface_hub.utils import get_session, hf_raise_for_status, HfHubHTTPError
    
            response = get_session().post(...)
            try:
                hf_raise_for_status(response)
            except HfHubHTTPError as e:
                print(str(e)) # formatted message
                e.request_id, e.server_message # details returned by server
    
                # Complete the error message with additional information once it's raised
                e.append_to_message("\n`create_commit` expects the repository to exist.")
                raise
        ```
    
        Args:
            response (`Response`):
                Response from the server.
            endpoint_name (`str`, *optional*):
                Name of the endpoint that has been called. If provided, the error message
                will be more complete.
    
        > [!WARNING]
        > Raises when the request has failed:
        >
        >     - [`~utils.RepositoryNotFoundError`]
        >         If the repository to download from cannot be found. This may be because it
        >         doesn't exist, because `repo_type` is not set correctly, or because the repo
        >         is `private` and you do not have access.
        >     - [`~utils.GatedRepoError`]
        >         If the repository exists but is gated and the user is not on the authorized
        >         list.
        >     - [`~utils.RevisionNotFoundError`]
        >         If the repository exists but the revision couldn't be find.
        >     - [`~utils.EntryNotFoundError`]
        >         If the repository exists but the entry (e.g. the requested file) couldn't be
        >         find.
        >     - [`~utils.BadRequestError`]
        >         If request failed with a HTTP 400 BadRequest error.
        >     - [`~utils.HfHubHTTPError`]
        >         If request failed for a reason not listed above.
        """
        try:
>           response.raise_for_status()

.venv/lib/python3.11/site-packages/huggingface_hub/utils/_http.py:402: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Response [404]>

    def raise_for_status(self):
        """Raises :class:`HTTPError`, if one occurred."""
    
        http_error_msg = ""
        if isinstance(self.reason, bytes):
            # We attempt to decode utf-8 first because some servers
            # choose to localize their reason strings. If the string
            # isn't utf-8, we fall back to iso-8859-1 for all other
            # encodings. (See PR #3538)
            try:
                reason = self.reason.decode("utf-8")
            except UnicodeDecodeError:
                reason = self.reason.decode("iso-8859-1")
        else:
            reason = self.reason
    
        if 400 <= self.status_code < 500:
            http_error_msg = (
                f"{self.status_code} Client Error: {reason} for url: {self.url}"
            )
    
        elif 500 <= self.status_code < 600:
            http_error_msg = (
                f"{self.status_code} Server Error: {reason} for url: {self.url}"
            )
    
        if http_error_msg:
>           raise HTTPError(http_error_msg, response=self)
E           requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://huggingface.co/api/models/gs://marin-us-central2/checkpoints/tootsie-8b-giraffe-32k-293fef/hf/step-2999

.venv/lib/python3.11/site-packages/requests/models.py:1026: HTTPError

The above exception was the direct cause of the following exception:

    def test_load_tokenizer_gcs():
        # sort of like a gs:// path insasmuch as it uses fsspec machinery
>       tokenizer = load_tokenizer("gs://marin-us-central2/checkpoints/tootsie-8b-giraffe-32k-293fef/hf/step-2999")
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

lib/levanter/tests/test_hf_utils.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
lib/levanter/src/levanter/compat/hf_checkpoints.py:1092: in load_tokenizer
    return AutoTokenizer.from_pretrained(
.venv/lib/python3.11/site-packages/transformers/models/auto/tokenization_auto.py:1156: in from_pretrained
    return tokenizer_class.from_pretrained(pretrained_model_name_or_path, *inputs, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:2113: in from_pretrained
    return cls._from_pretrained(
.venv/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:2395: in _from_pretrained
    tokenizer = cls._patch_mistral_regex(
.venv/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:2438: in _patch_mistral_regex
    if _is_local or is_base_mistral(pretrained_model_name_or_path):
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:2432: in is_base_mistral
    model = model_info(model_id)
            ^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py:114: in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/huggingface_hub/hf_api.py:2661: in model_info
    hf_raise_for_status(r)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

response = <Response [404]>, endpoint_name = None

    def hf_raise_for_status(response: Response, endpoint_name: Optional[str] = None) -> None:
        """
        Internal version of `response.raise_for_status()` that will refine a
        potential HTTPError. Raised exception will be an instance of `HfHubHTTPError`.
    
        This helper is meant to be the unique method to raise_for_status when making a call
        to the Hugging Face Hub.
    
    
        Example:
        ```py
            import requests
            from huggingface_hub.utils import get_session, hf_raise_for_status, HfHubHTTPError
    
            response = get_session().post(...)
            try:
                hf_raise_for_status(response)
            except HfHubHTTPError as e:
                print(str(e)) # formatted message
                e.request_id, e.server_message # details returned by server
    
                # Complete the error message with additional information once it's raised
                e.append_to_message("\n`create_commit` expects the repository to exist.")
                raise
        ```
    
        Args:
            response (`Response`):
                Response from the server.
            endpoint_name (`str`, *optional*):
                Name of the endpoint that has been called. If provided, the error message
                will be more complete.
    
        > [!WARNING]
        > Raises when the request has failed:
        >
        >     - [`~utils.RepositoryNotFoundError`]
        >         If the repository to download from cannot be found. This may be because it
        >         doesn't exist, because `repo_type` is not set correctly, or because the repo
        >         is `private` and you do not have access.
        >     - [`~utils.GatedRepoError`]
        >         If the repository exists but is gated and the user is not on the authorized
        >         list.
        >     - [`~utils.RevisionNotFoundError`]
        >         If the repository exists but the revision couldn't be find.
        >     - [`~utils.EntryNotFoundError`]
        >         If the repository exists but the entry (e.g. the requested file) couldn't be
        >         find.
        >     - [`~utils.BadRequestError`]
        >         If request failed with a HTTP 400 BadRequest error.
        >     - [`~utils.HfHubHTTPError`]
        >         If request failed for a reason not listed above.
        """
        try:
            response.raise_for_status()
        except HTTPError as e:
            error_code = response.headers.get("X-Error-Code")
            error_message = response.headers.get("X-Error-Message")
    
            if error_code == "RevisionNotFound":
                message = f"{response.status_code} Client Error." + "\n\n" + f"Revision Not Found for url: {response.url}."
                raise _format(RevisionNotFoundError, message, response) from e
    
            elif error_code == "EntryNotFound":
                message = f"{response.status_code} Client Error." + "\n\n" + f"Entry Not Found for url: {response.url}."
                raise _format(EntryNotFoundError, message, response) from e
    
            elif error_code == "GatedRepo":
                message = (
                    f"{response.status_code} Client Error." + "\n\n" + f"Cannot access gated repo for url {response.url}."
                )
                raise _format(GatedRepoError, message, response) from e
    
            elif error_message == "Access to this resource is disabled.":
                message = (
                    f"{response.status_code} Client Error."
                    + "\n\n"
                    + f"Cannot access repository for url {response.url}."
                    + "\n"
                    + "Access to this resource is disabled."
                )
                raise _format(DisabledRepoError, message, response) from e
    
            elif error_code == "RepoNotFound" or (
                response.status_code == 401
                and error_message != "Invalid credentials in Authorization header"
                and response.request is not None
                and response.request.url is not None
                and REPO_API_REGEX.search(response.request.url) is not None
            ):
                # 401 is misleading as it is returned for:
                #    - private and gated repos if user is not authenticated
                #    - missing repos
                # => for now, we process them as `RepoNotFound` anyway.
                # See https://gist.github.com/Wauplin/46c27ad266b15998ce56a6603796f0b9
                message = (
                    f"{response.status_code} Client Error."
                    + "\n\n"
                    + f"Repository Not Found for url: {response.url}."
                    + "\nPlease make sure you specified the correct `repo_id` and"
                    " `repo_type`.\nIf you are trying to access a private or gated repo,"
                    " make sure you are authenticated. For more details, see"
                    " https://huggingface.co/docs/huggingface_hub/authentication"
                )
                raise _format(RepositoryNotFoundError, message, response) from e
    
            elif response.status_code == 400:
                message = (
                    f"\n\nBad request for {endpoint_name} endpoint:" if endpoint_name is not None else "\n\nBad request:"
                )
                raise _format(BadRequestError, message, response) from e
    
            elif response.status_code == 403:
                message = (
                    f"\n\n{response.status_code} Forbidden: {error_message}."
                    + f"\nCannot access content at: {response.url}."
                    + "\nMake sure your token has the correct permissions."
                )
                raise _format(HfHubHTTPError, message, response) from e
    
            elif response.status_code == 416:
                range_header = response.request.headers.get("Range")
                message = f"{e}. Requested range: {range_header}. Content-Range: {response.headers.get('Content-Range')}."
                raise _format(HfHubHTTPError, message, response) from e
    
            # Convert `HTTPError` into a `HfHubHTTPError` to display request information
            # as well (request id and/or server error message)
>           raise _format(HfHubHTTPError, str(e), response) from e
E           huggingface_hub.errors.HfHubHTTPError: 404 Client Error: Not Found for url: https://huggingface.co/api/models/gs://marin-us-central2/checkpoints/tootsie-8b-giraffe-32k-293fef/hf/step-2999 (Request ID: Root=1-69433dbc-1bd6442a281797546dcc6d3f;ac2f2831-a974-4170-9b99-231e47950277)
E           
E           Sorry, we can't find the page you are looking for.

.venv/lib/python3.11/site-packages/huggingface_hub/utils/_http.py:475: HfHubHTTPError

Process finished with exit code 1
```


**Expected behavior**

Should pass


**Additional context**

Looks like they are hotpatching mistral tokenizers and we got caught in the crossfire. sigh

--- a/lib/levanter/src/levanter/compat/hf_checkpoints.py
+++ b/lib/levanter/src/levanter/compat/hf_checkpoints.py
@@ -35,7 +35,7 @@ from haliax._src.state_dict import flatten_modules_for_export, to_state_dict
 from haliax.jax_utils import is_jax_array_like
 from haliax.partitioning import ResourceMapping
 from haliax.state_dict import StateDict, from_torch_compatible_state_dict, save_state_dict
-from huggingface_hub import HfApi, hf_hub_download, repo_exists, snapshot_download
+from huggingface_hub import HfApi, hf_hub_download, repo_exists, snapshot_download, ModelInfo
 from huggingface_hub.file_download import repo_folder_name
 from huggingface_hub.utils import EntryNotFoundError, GatedRepoError, HFValidationError
 from jax import ShapeDtypeStruct
@@ -1323,9 +1323,21 @@ def _patch_hf_hub_download():
 
         huggingface_hub.utils._validators.validate_repo_id = custom_validate_repo_id
 
+        # transformers calls  model_info in _patch_mistral_regex to check if model is a base Mistral model
+        original_model_info = huggingface_hub.hf_api.model_info
+
+        def custom_model_info(repo_id, *args, **kwargs) -> ModelInfo:
+            if _is_url_like(repo_id):
+                # `tags=None` makes is_base_mistral return False, skipping the problematic code path
+                return ModelInfo(id="monkeypatched", tags=None)
+            return original_model_info(repo_id, *args, **kwargs)
+
+        huggingface_hub.hf_api.model_info = custom_model_info
+
         try:
             yield custom_hf_hub_download
         finally:
             # Restore the original implementation
             transformers.utils.hub.hf_hub_download = original_hf_hub_download
             huggingface_hub.utils._validators.validate_repo_id = original_validate_repo_id
+            huggingface_hub.hf_api.model_info = original_model_info

--- a/lib/levanter/tests/test_hf_utils.py
+++ b/lib/levanter/tests/test_hf_utils.py
@@ -5,7 +5,7 @@ import os
 
 from fsspec import AbstractFileSystem
 
-from levanter.compat.hf_checkpoints import load_tokenizer
+from levanter.compat.hf_checkpoints import _patch_hf_hub_download, load_tokenizer
 from levanter.utils.hf_utils import byte_length_of_token
 from test_utils import skip_if_hf_model_not_accessible
 
@@ -28,6 +28,17 @@ def test_load_tokenizer_in_memory_fs():
         )
     tokenizer = load_tokenizer("memory://foo/")
     assert len(tokenizer) == 5027
+
+
+def test_model_info_patch_for_fsspec_urls():
+    """transformers calls model_info() in _patch_mistral_regex to check if a model is a base Mistral model."""
+    import huggingface_hub
+
+    with _patch_hf_hub_download():
+        # This should NOT raise or make a network call - it should return a mock
+        result = huggingface_hub.hf_api.model_info("memory://some/path")
+        assert result.id == "monkeypatched"
+        assert result.tags is None
 
 
 @skip_if_hf_model_not_accessible("NousResearch/Llama-2-7b-hf")

